src/SourceBuild/content/eng/tools/BinaryToolKit/BinaryTool.DetectBinaries.cs

```diff
@@ -0,0 +1,110 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using Microsoft.Extensions.FileSystemGlobbing;
+using Microsoft.Extensions.Logging;
+
+namespace BinaryToolKit;
+
+public partial class BinaryTool
+{
+    private const string Utf16Marker = "UTF-16";
+    private const int ChunkSize = 4096;
+
+    private async Task<IEnumerable<string>> DetectBinariesAsync(string targetDirectory)
+    {
+        Log.LogInformation($"Detecting binaries in {targetDirectory}...");
+
+        var matcher = new Matcher(StringComparison.Ordinal);
+        matcher.AddInclude("**/*");
+        matcher.AddExcludePatterns(new[]
+        {
+            "**/.dotnet/**",
+            "**/.git/**",
+            "**/git-info/**",
+            "**/artifacts/**",
+            "**/prereqs/packages/**",
+            "**/.packages/**"
+        });
+
+        IEnumerable<string> matchingFiles = matcher.GetResultsInFullPath(targetDirectory);
+
+        var tasks = matchingFiles
+            .Select(async file =>
+            {
+                return await IsBinary(file) ? file : null;
+            });
+
+        var binaryFiles = (await Task.WhenAll(tasks)).OfType<string>().Select(file => file.Substring(targetDirectory.Length + 1));
```

Comment:
The OfType() was suggested in [this comment](https://github.com/dotnet/installer/pull/18726#discussion_r1503983129). It is a cleaner way of filtering out any null tasks.

Reply:
> Consider moving the file substring into the task. e.g. run as much logic as possible in parallel.

From what I can find, there's not a great way to do this. Using Task.WhenAny() would work, but it can be slow when there are many tasks. An alternative is to do what's described in https://devblogs.microsoft.com/pfxteam/processing-tasks-as-they-complete/, but the perf analysis doesn't show a big difference for this tool.

Reply:
All I meant was you could move this string manipulation logic into the previous Select.

return await IsBinary(file) ? file.Substring(targetDirectory.Length + 1) : null;